<div style="background-color:#EAF4EC; padding:20px; border-radius:10px;">

<h2 style="color:#2F6F4E; text-align:center; margin-bottom:5px;">
Forecasting: Econometric Baseline — Fixed Effects Panel Regression
</h2>

<h4 style="color:#2F6F4E; text-align:center; margin-top:0;">
Master Thesis – ESG Governance Indicators (EU-27)
</h4>

<p style="font-size:14px; color:#2F6F4E;">
This notebook implements the <strong>econometric baseline</strong> for the forecasting stage of the
CRISP-ML(Q) methodology. The objective is to estimate a <strong>Fixed Effects panel regression model</strong>
for each ESG governance indicator across the EU-27.
</p>

<p style="font-size:14px; color:#2F6F4E;">
The results obtained in this notebook serve as a <strong>reference benchmark</strong> against which the performance
of non-linear machine learning models (e.g., Random Forest and XGBoost) is assessed in later stages
of the analysis.
</p>

</div>

<div style="background-color:#EAF4EC; padding:16px; border-radius:10px;"> <h2 style="color:#2F6F4E; margin-bottom:5px; font-size:20px;"> 1. Data Integrity — Do we have the dataset we expect? </h2> </div>

In [1]:
import pandas as pd
import numpy as np
import re
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.metrics import r2_score
from linearmodels.panel import PanelOLS, PooledOLS, RandomEffects
import statsmodels.api as sm

In [2]:

df_wide = pd.read_csv("../data/processed/data_final_EU27.csv")

In [3]:
print(df_wide.shape)
df_wide.head()

(405, 28)


,Country Name,Country Code,Indicator Name,Indicator Code,2000,2001,2002,2003,2004,2005,...,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023
0,Austria,AUT,Control of Corruption: Estimate,CC.EST,1.751059,1.751059,1.915159,1.963869,2.026624,1.912423,...,1.466502,1.472044,1.496881,1.500820,1.568587,1.521324,1.477709,1.242727,1.258587,1.133653
1,Austria,AUT,GDP growth (annual %),NY.GDP.MKTP.KD.ZG,3.189523,1.316987,1.484369,1.141565,2.565255,2.320365,...,0.755799,1.303523,2.117220,2.272250,2.484221,1.754976,-6.318255,4.795329,5.277894,-0.954963
2,Austria,AUT,Government Effectiveness: Estimate,GE.EST,1.847265,1.847265,1.880228,1.930461,1.843771,1.679529,...,1.552996,1.450979,1.474811,1.467752,1.456117,1.492126,1.605863,1.528295,1.464923,1.326681
3,Austria,AUT,Individuals using the Internet (% of population),IT.NET.USER.ZS,33.700000,39.200000,36.600000,42.700000,54.300000,58.000000,...,81.000000,83.900000,84.300000,87.900000,87.500000,87.800000,87.500000,92.500000,93.600000,95.300000
4,Austria,AUT,Net migration,SM.POP.NETM,17631.000000,39866.000000,37508.000000,44362.000000,55007.000000,50676.000000,...,74375.000000,115175.000000,65976.000000,45478.000000,35296.000000,41090.000000,39601.000000,52150.000000,157365.000000,4819.000000


In [4]:
# % de NaNs per column  
nan_pct = df_wide.isnull().mean().sort_values(ascending=False) * 100

print("NaNs por coluna (%)")
print(nan_pct[nan_pct > 0].round(2))

print(f"\n Shape original: {df_wide.shape} ")
print(f"Linhas sem nenhum NaN: {df_wide.dropna().shape[0]} ")
print(f"Linhas perdidas com dropna: {df_wide.shape[0] - df_wide.dropna().shape[0]} ")

NaNs por coluna (%)
Series([], dtype: float64)

 Shape original: (405, 28) 
Linhas sem nenhum NaN: 405 
Linhas perdidas com dropna: 0 


<div style="background-color:#EAF4EC; padding:16px; border-radius:10px;"> <h2 style="color:#2F6F4E; margin-bottom:5px; font-size:20px;"> 2. Temporal Structure — Identifying the available time dimension </h2> </div>

In [5]:
year_cols = [c for c in df_wide.columns if str(c).isdigit()]
print("Years:", year_cols[:5], "...", year_cols[-5:])
print("N years:", len(year_cols))


Years: ['2000', '2001', '2002', '2003', '2004'] ... ['2019', '2020', '2021', '2022', '2023']
N years: 24


<div style="background-color:#EAF4EC; padding:16px; border-radius:10px;"> <h2 style="color:#2F6F4E; margin-bottom:5px; font-size:20px;"> Panel Structure — Constructing a lagged country–year panel </h2> </div>

In [6]:
def prepare_panel_with_covariates(df_wide_raw: pd.DataFrame, target_code: str) -> pd.DataFrame:
    
    id_cols = ["Country Name", "Country Code", "Indicator Name", "Indicator Code"]
    
    # Create long format from the wide dataset
    df_long = df_wide_raw.melt(
        id_vars=id_cols,
        var_name="Year",
        value_name="Value"
    )
    
    df_long["Year"] = df_long["Year"].astype(int)
    
    # Pivot: rows = country-year, columns = indicators
    df_p = df_long.pivot_table(
        index=["Country Code", "Year"],
        columns="Indicator Code",
        values="Value"
    ).sort_index()
    
    # Target variable
    df_p["y"] = df_p[target_code]
    
    # Covariates = all indicators except the target
    covariates = [c for c in df_p.columns if c not in [target_code, "y"]]
    
    # Lagged target
    df_p["y_lag1"] = df_p.groupby("Country Code")[target_code].shift(1)
    
    # Lagged covariates
    for col in covariates:
        df_p[f"{col}_lag1"] = df_p.groupby("Country Code")[col].shift(1)
    
    # Keep only y and lagged variables
    lag_cols = ["y", "y_lag1"] + [f"{c}_lag1" for c in covariates]
    
    return df_p[lag_cols].dropna()

<div style="background-color:#EAF4EC; padding:16px; border-radius:10px;"> <h2 style="color:#2F6F4E; margin-bottom:5px; font-size:20px;"> 6. Temporal Splitting  </h2> </div>

In [7]:
def time_split(panel_df: pd.DataFrame, n_train_years=16, n_val_years=3, n_test_years=2):
    years = sorted(panel_df.index.get_level_values("Year").unique())
    total_needed = n_train_years + n_val_years + n_test_years

    if len(years) < total_needed:
        raise ValueError(
            f"Not enough years to split: have {len(years)} years, need {total_needed}."
        )

    test_years = years[-n_test_years:]
    val_years = years[-(n_test_years + n_val_years):-n_test_years]
    train_years = years[-(n_test_years + n_val_years + n_train_years):-(n_test_years + n_val_years)]

    train_df = panel_df.loc[(slice(None), train_years), :]
    val_df   = panel_df.loc[(slice(None), val_years), :]
    test_df  = panel_df.loc[(slice(None), test_years), :]

    return train_df, val_df, test_df, train_years, val_years, test_years


In [8]:
panel_cc = prepare_panel_with_covariates(df_wide, "CC.EST")  

train_df, val_df, test_df, train_years, val_years, test_years = time_split(panel_cc)

print("Train years:", train_years[0], "-", train_years[-1], "| rows:", len(train_df))
print("Val years:", val_years[0], "-", val_years[-1], "| rows:", len(val_df))
print("Test years:", test_years[0], "-", test_years[-1], "| rows:", len(test_df))

Train years: 2003 - 2018 | rows: 432
Val years: 2019 - 2021 | rows: 81
Test years: 2022 - 2023 | rows: 54


<div style="background-color:#EAF4EC; padding:16px; border-radius:10px;"> <h2 style="color:#2F6F4E; margin-bottom:5px; font-size:20px;"> 7. Econometric Baseline — Fixed Effects panel regression </h2> </div>

In [9]:
def fit_fe(train_df: pd.DataFrame):
    X_train = train_df.drop(columns=["y"]).copy()
    X_train["const"] = 1.0
    y_train = train_df["y"]
    model = PanelOLS(y_train, X_train, entity_effects=True)
    res = model.fit(cov_type="clustered", cluster_entity=True)
    return res

def predict_and_metrics(res, df_part: pd.DataFrame):
    X = df_part.drop(columns=["y"]).copy()
    X["const"] = 1.0
    y_true = df_part["y"]

    train_cols = res.model.exog.vars
    X = X[train_cols]

    y_pred = res.predict(exog=X).iloc[:, 0]

    mae  = mean_absolute_error(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    r2   = r2_score(y_true, y_pred)

    return mae, rmse, r2

<div style="background-color:#EAF4EC; padding:16px; border-radius:10px;"> <h2 style="color:#2F6F4E; margin-bottom:5px; font-size:20px;"> 8. Baseline Evaluation — Loop over all indicators </h2> </div>

In [10]:
BASELINE_INDICATORS = [
     "CC.EST",
    "GB.XPD.RSDV.GD.ZS",
    "GE.EST",
    "IP.JRN.ARTC.SC",
    "IP.PAT.RESD",
    "IT.NET.USER.ZS",
    "NY.GDP.MKTP.KD.ZG",
    "PV.EST",
    "RL.EST",
    "RQ.EST",
    "SE.ENR.PRSC.FM.ZS",
    "SG.GEN.PARL.ZS",
    "SL.TLF.CACT.FM.ZS",
    "SM.POP.NETM",
    "VA.EST"
]

In [11]:
def prepare_panel_with_covariates(df_wide_raw: pd.DataFrame, target_code: str) -> pd.DataFrame:
    id_cols = ["Country Name", "Country Code", "Indicator Name", "Indicator Code"]
    year_cols = [c for c in df_wide_raw.columns if c not in id_cols]

    df = df_wide_raw.set_index(id_cols)[year_cols]
    df = df.stack().reset_index()
    df.columns = ["Country Name", "Country Code", "Indicator Name", "Indicator Code", "Year", "Value"]
    df["Year"] = df["Year"].astype(int)

    df_p = df.pivot_table(
        index=["Country Code", "Year"],
        columns="Indicator Code",
        values="Value"
    ).sort_index()

    df_p["y"] = df_p[target_code]
    covariates = [c for c in df_p.columns if c != target_code and c != "y"]

    df_p["y_lag1"] = df_p.groupby("Country Code")[target_code].shift(1)
    for col in covariates:
        df_p[f"{col}_lag1"] = df_p.groupby("Country Code")[col].shift(1)

    lag_cols = ["y", "y_lag1"] + [f"{c}_lag1" for c in covariates]
    return df_p[lag_cols].dropna()

In [12]:
def time_split(panel_df: pd.DataFrame, n_train_years=16, n_val_years=3, n_test_years=2):
    years = sorted(panel_df.index.get_level_values("Year").unique())
    total_needed = n_train_years + n_val_years + n_test_years

    if len(years) < total_needed:
        raise ValueError(f"Not enough years to split: have {len(years)}, need {total_needed}.")

    test_years  = years[-n_test_years:]
    val_years   = years[-(n_test_years + n_val_years):-n_test_years]
    train_years = years[-(n_test_years + n_val_years + n_train_years):-(n_test_years + n_val_years)]

    train_df = panel_df.loc[(slice(None), train_years), :]
    val_df   = panel_df.loc[(slice(None), val_years), :]
    test_df  = panel_df.loc[(slice(None), test_years), :]

    return train_df, val_df, test_df, train_years, val_years, test_years

In [13]:
def fit_fe(train_df: pd.DataFrame):
    X_train = train_df.drop(columns=["y"]).copy()
    X_train["const"] = 1.0
    y_train = train_df["y"]

    model = PanelOLS(y_train, X_train, entity_effects=True, drop_absorbed=True)
    res = model.fit(cov_type="clustered", cluster_entity=True)
    return res

In [14]:
baseline_results = []
baseline_models  = {}

for ind in BASELINE_INDICATORS:
    panel_df = prepare_panel_with_covariates(df_wide, ind)

    train_df, val_df, test_df, train_years, val_years, test_years = time_split(panel_df)

    res = fit_fe(train_df)

    # Metrics
    mae_train, rmse_train, r2_train = predict_and_metrics(res, train_df)
    mae_val,   rmse_val,   r2_val   = predict_and_metrics(res, val_df)
    mae_test,  rmse_test,  r2_test  = predict_and_metrics(res, test_df)

    baseline_models[ind] = res

    baseline_results.append({
        "Indicator Code": ind,

        "MAE_Train":   mae_train,
        "RMSE_Train":  rmse_train,
        "R2_Train":    r2_train,

        "MAE_Val":     mae_val,
        "RMSE_Val":    rmse_val,
        "R2_Val":      r2_val,

        "MAE_Test":    mae_test,
        "RMSE_Test":   rmse_test,
        "R2_Test":     r2_test,

        "Train_Years": f"{train_years[0]}-{train_years[-1]}",
        "Val_Years":   f"{val_years[0]}-{val_years[-1]}",
        "Test_Years":  f"{test_years[0]}-{test_years[-1]}",

        "N_Train":     len(train_df),
        "N_Val":       len(val_df),
        "N_Test":      len(test_df),
        "N_Total":     len(train_df) + len(val_df) + len(test_df),

        "Countries":   panel_df.index.get_level_values("Country Code").nunique()
    })

    print(
        f"{ind} | "
        f"MAE_Train: {mae_train:.4f} | RMSE_Train: {rmse_train:.4f} | R2_Train: {r2_train:.4f} | "
        f"MAE_Val: {mae_val:.4f} | RMSE_Val: {rmse_val:.4f} | R2_Val: {r2_val:.4f} | "
        f"MAE_Test: {mae_test:.4f} | RMSE_Test: {rmse_test:.4f} | R2_Test: {r2_test:.4f}"
    )

baseline_results_df = pd.DataFrame(baseline_results)
baseline_results_df

CC.EST | MAE_Train: 0.2210 | RMSE_Train: 0.3049 | R2_Train: 0.8508 | MAE_Val: 0.2100 | RMSE_Val: 0.2941 | R2_Val: 0.8504 | MAE_Test: 0.1889 | RMSE_Test: 0.2723 | R2_Test: 0.8672
GB.XPD.RSDV.GD.ZS | MAE_Train: 0.2319 | RMSE_Train: 0.3532 | R2_Train: 0.8424 | MAE_Val: 0.2480 | RMSE_Val: 0.3612 | R2_Val: 0.8349 | MAE_Test: 0.2397 | RMSE_Test: 0.3365 | R2_Test: 0.8599
GE.EST | MAE_Train: 0.2349 | RMSE_Train: 0.3366 | R2_Train: 0.6864 | MAE_Val: 0.2418 | RMSE_Val: 0.3412 | R2_Val: 0.6375 | MAE_Test: 0.2346 | RMSE_Test: 0.3161 | R2_Test: 0.6836
IP.JRN.ARTC.SC | MAE_Train: 1082.1865 | RMSE_Train: 1766.2072 | R2_Train: 0.9945 | MAE_Val: 1565.9368 | RMSE_Val: 2907.4087 | R2_Val: 0.9895 | MAE_Test: 1019.7181 | RMSE_Test: 1832.2609 | R2_Test: 0.9959
IP.PAT.RESD | MAE_Train: 1477.1524 | RMSE_Train: 2940.7982 | R2_Train: 0.8979 | MAE_Val: 1430.8541 | RMSE_Val: 2461.2812 | R2_Val: 0.9136 | MAE_Test: 1440.9355 | RMSE_Test: 2518.4945 | R2_Test: 0.8966
IT.NET.USER.ZS | MAE_Train: 3.6432 | RMSE_Train: 6

,Indicator Code,MAE_Train,RMSE_Train,R2_Train,MAE_Val,RMSE_Val,R2_Val,MAE_Test,RMSE_Test,R2_Test,Train_Years,Val_Years,Test_Years,N_Train,N_Val,N_Test,N_Total,Countries
0,CC.EST,0.220966,0.304922,0.850849,0.210008,0.294055,0.850412,0.188885,0.272289,0.867228,2003-2018,2019-2021,2022-2023,432,81,54,567,27
1,GB.XPD.RSDV.GD.ZS,0.231940,0.353153,0.842390,0.248011,0.361163,0.834889,0.239747,0.336504,0.859864,2003-2018,2019-2021,2022-2023,432,81,54,567,27
2,GE.EST,0.234939,0.336604,0.686414,0.241778,0.341174,0.637508,0.234583,0.316088,0.683594,2003-2018,2019-2021,2022-2023,432,81,54,567,27
3,IP.JRN.ARTC.SC,1082.186460,1766.207192,0.994523,1565.936781,2907.408657,0.989532,1019.718146,1832.260946,0.995862,2003-2018,2019-2021,2022-2023,432,81,54,567,27
4,IP.PAT.RESD,1477.152369,2940.798213,0.897940,1430.854065,2461.281247,0.913627,1440.935464,2518.494501,0.896625,2003-2018,2019-2021,2022-2023,432,81,54,567,27
5,IT.NET.USER.ZS,3.643180,6.550123,0.887156,4.198881,6.385350,0.224824,3.661270,5.451239,-0.247834,2003-2018,2019-2021,2022-2023,432,81,54,567,27
6,NY.GDP.MKTP.KD.ZG,6.426929,12.291501,-9.465222,7.571347,12.280712,-4.045107,6.468988,10.155664,-12.918698,2003-2018,2019-2021,2022-2023,432,81,54,567,27
7,PV.EST,0.221615,0.277261,0.522948,0.187454,0.240996,0.159176,0.165829,0.225656,-0.090167,2003-2018,2019-2021,2022-2023,432,81,54,567,27
8,RL.EST,0.162881,0.311260,0.742397,0.157006,0.287697,0.748604,0.143490,0.254486,0.795642,2003-2018,2019-2021,2022-2023,432,81,54,567,27
9,RQ.EST,0.141464,0.267034,0.617778,0.161224,0.268308,0.671634,0.143365,0.216542,0.812630,2003-2018,2019-2021,2022-2023,432,81,54,567,27


# Baseline Panel Regression — Results Analysis

## Temporal Split Configuration
| Period | Years | Observations |
|---|---|---|
| **Train** | 2003–2018 | 432 |
| **Validation** | 2019–2021 | 81 |
| **Test** | 2022–2023 | 54 |

> Balanced panel with **27 countries** and **15 indicators**. Year 2000 is sacrificed for lag construction. The model uses country Fixed Effects with clustered standard errors.

---

## Results by Group

### Group 1 — Strong Performance (R²_Test > 0.85)

| Indicator Code | MAE_Test | RMSE_Test | R²_Test |
|---|---|---|---|
| IP.JRN.ARTC.SC | 1019.72 | 1832.26 | **0.9959** |
| IP.PAT.RESD | 1440.94 | 2518.49 | **0.8966** |
| SL.TLF.CACT.FM.ZS | 1.31 | 1.73 | **0.8975** |
| CC.EST | 0.19 | 0.27 | **0.8672** |
| GB.XPD.RSDV.GD.ZS | 0.24 | 0.34 | **0.8599** |

The model captures the temporal dynamics of these indicators well. Typical of **persistent** series — indicators that change slowly year-on-year, where the lagged value is a strong predictor.

---

### Group 2 — Moderate Performance (R²_Test 0.65–0.85)

| Indicator Code | MAE_Test | RMSE_Test | R²_Test |
|---|---|---|---|
| RQ.EST | 0.14 | 0.22 | **0.8126** |
| RL.EST | 0.14 | 0.25 | **0.7956** |
| SG.GEN.PARL.ZS | 3.23 | 4.24 | **0.7890** |
| GE.EST | 0.23 | 0.32 | **0.6836** |
| VA.EST | 0.12 | 0.21 | **0.6789** |

Acceptable baseline performance. Room for improvement with more sophisticated forecasting models.

---

### Group 3 — Poor Performance (R²_Test < 0)

| Indicator Code | MAE_Test | RMSE_Test | R²_Test |
|---|---|---|---|
| NY.GDP.MKTP.KD.ZG | 6.47 | 10.16 | **-12.9187** |
| SM.POP.NETM | 260206.13 | 468510.18 | **-4.2826** |
| SE.ENR.PRSC.FM.ZS | 0.03 | 0.04 | **-1.6092** |
| IT.NET.USER.ZS | 3.66 | 5.45 | **-0.2478** |
| PV.EST | 0.17 | 0.23 | **-0.0902** |

A negative R² means the model performs **worse than simply predicting the mean**. Each case has a structural explanation:
- **NY.GDP.MKTP.KD.ZG** — GDP growth is highly volatile and sensitive to macroeconomic shocks not captured by lagged values
- **SM.POP.NETM** — net migration is driven by sudden geopolitical and economic events
- **SE.ENR.PRSC.FM.ZS** — school enrollment shows non-linear trends across EU countries
- **IT.NET.USER.ZS** — internet usage was in a non-linear adoption phase during the sample period
- **PV.EST** — political stability is influenced by discrete events that lags cannot anticipate

---

## Key Takeaways

1. **Fixed Effects performs well for slow-moving, persistent indicators** — governance, patents, R&D expenditure, and labour market indicators.
2. **The model struggles with volatile and event-driven indicators** — GDP growth, migration, and internet adoption.
3. **Negative R² results justify the use of more advanced models** (Random Forest, XGBoost), providing a clear motivation for moving beyond the baseline.
4. **The panel structure is balanced** — all 15 indicators cover the same 27 EU countries over the same time periods.